File: v4/rust/iprange-livedb/src/recovery/structure_index.rs
//! Recovery of authoritative structured payloads by source-local ID.

use std::marker::PhantomData;

use crate::cancellation::CancellationToken;
use crate::contract::MetaV4;
use crate::error::{Error, Result};
use crate::mapping::{ByteSource, Mapping, PageView};
use crate::structured_value::codec::{self, PayloadCodec};
use crate::structured_value::{payload_digest, table};
use crate::validation::{ValidationObject, ValidationReason};

use super::membership_table::MembershipIndex;
use super::page_set::{PageClaim, PageSet};
use super::report::{emit_page_unknown as emit, RecoverySink, Reporter};
use super::structure_table::{Insert, Locator, StructureIndex};
use super::tables::Tables;

pub(crate) fn count<P: PayloadCodec>(
    mapping: &Mapping,
    meta: MetaV4,
    pages: &mut PageSet,
    cancellation: &CancellationToken,
) -> Result<u64> {
    let mut events = Counter::<P> {
        meta,
        count: 0,
        marker: PhantomData,
    };
    scan::<P, _>(
        mapping,
        meta,
        meta.structure_id_root,
        pages,
        cancellation,
        &mut events,
    )?;
    Ok(events.count)
}

pub(crate) fn recover<P: PayloadCodec, S: RecoverySink>(
    mapping: &Mapping,
    meta: MetaV4,
    memberships: &MembershipIndex,
    pages: &mut PageSet,
    tables: &mut Tables,
    cancellation: &CancellationToken,
    reporter: &mut Reporter<'_, S>,
) -> Result<StructureIndex> {
    if meta.structure_kind() != Some(P::KIND) {
        return Err(Error::UnsupportedStructure(meta.structure_kind_code));
    }
    let mut entries = StructureIndex::new(tables, P::KIND);
    {
        let mut events = Events::<P, S> {
            meta,
            reporter,
            memberships,
            entries: &mut entries,
            tables,
            marker: PhantomData,
        };
        scan::<P, _>(
            mapping,
            meta,
            meta.structure_id_root,
            pages,
            cancellation,
            &mut events,
        )?;
    }
    reconcile_ids(&entries, tables, reporter)?;
    report_outcomes(&entries, tables, reporter)?;
    Ok(entries)
}

trait TableEvents {
    fn page_accepted(&mut self) -> Result<()>;
    fn page_rejected(&mut self, io_unreadable: bool) -> Result<()>;
    fn unknown(&mut self, reason: ValidationReason, page: Option<u32>) -> Result<()>;
    fn leaf<B: ByteSource>(&mut self, page: u32, expected_id: u64, cell: B) -> Result<()>;
}

struct Counter<P> {
    meta: MetaV4,
    count: u64,
    marker: PhantomData<P>,
}

impl<P: PayloadCodec> TableEvents for Counter<P> {
    fn page_accepted(&mut self) -> Result<()> {
        Ok(())
    }

    fn page_rejected(&mut self, _io_unreadable: bool) -> Result<()> {
        Ok(())
    }

    fn unknown(&mut self, _reason: ValidationReason, _page: Option<u32>) -> Result<()> {
        Ok(())
    }

    fn leaf<B: ByteSource>(&mut self, _page: u32, expected_id: u64, cell: B) -> Result<()> {
        if codec::decode_record::<P, _>(cell).is_ok_and(|record| {
            u64::from(record.id) == expected_id
                && expected_id < self.meta.structure_id_limit
                && payload_digest::<P>(&record.payload).is_ok_and(|digest| digest == record.digest)
        }) {
            self.count = self
                .count
                .checked_add(1)
                .ok_or(Error::ArithmeticOverflow("recovery structure count"))?;
        }
        Ok(())
    }
}

struct Events<'a, 'b, 'c, P, S> {
    meta: MetaV4,
    reporter: &'a mut Reporter<'b, S>,
    memberships: &'a MembershipIndex,
    entries: &'a mut StructureIndex,
    tables: &'a mut Tables,
    marker: PhantomData<&'c P>,
}

impl<P: PayloadCodec, S: RecoverySink> TableEvents for Events<'_, '_, '_, P, S> {
    fn page_accepted(&mut self) -> Result<()> {
        self.reporter.page_accepted()
    }

    fn page_rejected(&mut self, io_unreadable: bool) -> Result<()> {
        self.reporter.page_rejected(io_unreadable)
    }

    fn unknown(&mut self, reason: ValidationReason, page: Option<u32>) -> Result<()> {
        emit(
            self.reporter,
            reason,
            ValidationObject::StructureDictionary,
            page,
        )
    }

    fn leaf<B: ByteSource>(&mut self, page: u32, expected_id: u64, cell: B) -> Result<()> {
        self.reporter.structure_examined()?;
        let Ok(record) = codec::decode_record::<P, _>(cell) else {
            return self.reporter.structure_rejected(1);
        };
        if u64::from(record.id) != expected_id || expected_id >= self.meta.structure_id_limit {
            self.reporter.structure_rejected(1)?;
            return emit(
                self.reporter,
                ValidationReason::StructureInvalid,
                ValidationObject::StructureDictionary,
                Some(page),
            );
        }
        if payload_digest::<P>(&record.payload)? != record.digest {
            self.reporter.structure_rejected(1)?;
            return emit(
                self.reporter,
                ValidationReason::StructureHashInvalid,
                ValidationObject::StructureDictionary,
                Some(page),
            );
        }
        let membership_id = P::membership_id(&record.payload);
        let rejected =
            membership_id != 0 && self.memberships.get(self.tables, membership_id)?.is_none();
        if rejected {
            emit(
                self.reporter,
                ValidationReason::StructureMembershipInvalid,
                ValidationObject::StructureDictionary,
                Some(page),
            )?;
        }
        self.entries.push(
            self.tables,
            Locator {
                id: record.id,
                membership_id,
                leaf_page: page,
                payload: record.payload,
                rejected,
            },
        )
    }
}

const MAX_DEPTH: usize = 4;

fn scan<P: PayloadCodec, E: TableEvents>(
    mapping: &Mapping,
    meta: MetaV4,
    root: u32,
    pages: &mut PageSet,
    cancellation: &CancellationToken,
    events: &mut E,
) -> Result<()> {
    if root == 0 {
        return Ok(());
    }
    let level = table::required_level::<P>(meta.structure_id_limit)?;
    let mut path = [0; MAX_DEPTH];
    scan_node::<P, E>(
        ScanContext {
            mapping,
            meta,
            pages,
            cancellation,
        },
        root,
        level,
        0,
        &mut path,
        0,
        events,
    )
}

struct ScanContext<'a> {
    mapping: &'a Mapping,
    meta: MetaV4,
    pages: &'a mut PageSet,
    cancellation: &'a CancellationToken,
}

#[allow(clippy::too_many_arguments)]
fn scan_node<P: PayloadCodec, E: TableEvents>(
    mut context: ScanContext<'_>,
    page_number: u32,
    expected_level: u16,
    base: u64,
    path: &mut [u32; MAX_DEPTH],
    depth: usize,
    events: &mut E,
) -> Result<()> {
    context.cancellation.check()?;
    if !claim_page(&mut context, page_number, path, depth, events)? {
        return Ok(());
    }
    let Some((page, header)) = read_page::<P, E>(&context, page_number, expected_level, events)?
    else {
        return Ok(());
    };
    if header.level == 0 {
        scan_leaf::<P, E>(&context, page, header, page_number, base, events)
    } else {
        scan_branch::<P, E>(
            context,
            page,
            header,
            page_number,
            base,
            path,
            depth,
            events,
        )
    }
}

fn claim_page<E: TableEvents>(
    context: &mut ScanContext<'_>,
    page_number: u32,
    path: &mut [u32; MAX_DEPTH],
    depth: usize,
    events: &mut E,
) -> Result<bool> {
    match context
        .pages
        .claim(page_number, context.meta.page_count, path, depth)?
    {
        PageClaim::Claimed => Ok(true),
        PageClaim::Rejected(reason) => {
            events.unknown(reason, Some(page_number))?;
            Ok(false)
        }
    }
}

fn read_page<'m, P: PayloadCodec, E: TableEvents>(
    context: &ScanContext<'m>,
    page_number: u32,
    expected_level: u16,
    events: &mut E,
) -> Result<Option<(PageView<'m>, table::Header)>> {
    let page =
        match super::page_read::checked(context.mapping, page_number, context.meta.page_count) {
            Ok(page) => page,
            Err(problem) => {
                reject_page(events, page_number, problem.reason, problem.io_unreadable)?;
                return Ok(None);
            }
        };
    let header =
        match table::inspect_header::<P, _>(page, context.meta.txn_id, Some(expected_level)) {
            Ok(header) => header,
            Err(problem) => {
                reject_page(events, page_number, header_reason(problem), false)?;
                return Ok(None);
            }
        };
    if !table::reserved_zero::<P, _>(page, header.level) {
        reject_page(
            events,
            page_number,
            ValidationReason::PageReservedNonzero,
            false,
        )?;
        return Ok(None);
    }
    events.page_accepted()?;
    Ok(Some((page, header)))
}

fn scan_leaf<P: PayloadCodec, E: TableEvents>(
    context: &ScanContext<'_>,
    page: PageView<'_>,
    header: table::Header,
    page_number: u32,
    base: u64,
    events: &mut E,
) -> Result<()> {
    let mut found = 0usize;
    for slot in 0..table::leaf_slots::<P>() {
        context.cancellation.check()?;
        let cell = table::leaf_cell::<P, _>(page, slot)?;
        if cell.all_zero(0, cell.len()) {
            continue;
        }
        found += 1;
        events.leaf(page_number, base + slot as u64, cell)?;
    }
    if found != header.item_count {
        events.unknown(ValidationReason::PageHeaderInvalid, Some(page_number))?;
    }
    Ok(())
}

#[allow(clippy::too_many_arguments)]
fn scan_branch<P: PayloadCodec, E: TableEvents>(
    context: ScanContext<'_>,
    page: PageView<'_>,
    header: table::Header,
    page_number: u32,
    base: u64,
    path: &mut [u32; MAX_DEPTH],
    depth: usize,
    events: &mut E,
) -> Result<()> {
    let span = table::coverage::<P>(header.level - 1)?;
    let mut found = 0usize;
    for index in 0..table::branch_slots() {
        context.cancellation.check()?;
        let child = table::raw_branch_child(page, index)?;
        if child == 0 {
            continue;
        }
        found += 1;
        if child < 2 || u64::from(child) >= context.meta.page_count {
            events.unknown(ValidationReason::PageOutOfBounds, Some(child))?;
            continue;
        }
        let child_base = base
            .checked_add(
                span.checked_mul(index as u64)
                    .ok_or(Error::ArithmeticOverflow("structure recovery coverage"))?,
            )
            .ok_or(Error::ArithmeticOverflow("structure recovery coverage"))?;
        scan_node::<P, E>(
            ScanContext {
                mapping: context.mapping,
                meta: context.meta,
                pages: context.pages,
                cancellation: context.cancellation,
            },
            child,
            header.level - 1,
            child_base,
            path,
            depth + 1,
            events,
        )?;
    }
    if found != header.item_count {
        events.unknown(ValidationReason::PageHeaderInvalid, Some(page_number))?;
    }
    Ok(())
}

fn reject_page<E: TableEvents>(
    events: &mut E,
    page_number: u32,
    reason: ValidationReason,
    io_unreadable: bool,
) -> Result<()> {
    events.page_rejected(io_unreadable)?;
    events.unknown(reason, Some(page_number))
}

fn header_reason(problem: table::HeaderProblem) -> ValidationReason {
    match problem {
        table::HeaderProblem::Header | table::HeaderProblem::Shape => {
            ValidationReason::PageHeaderInvalid
        }
        table::HeaderProblem::Born => ValidationReason::PageBornTxnInvalid,
        table::HeaderProblem::Type => ValidationReason::PageTypeMismatch,
        table::HeaderProblem::Level => ValidationReason::TreeLevelInvalid,
    }
}

fn reconcile_ids<S: RecoverySink>(
    entries: &StructureIndex,
    tables: &mut Tables,
    reporter: &mut Reporter<'_, S>,
) -> Result<()> {
    for index in 0..entries.records_len() {
        let entry = entries.record(tables, index)?;
        let Insert::Duplicate {
            first,
            newly_conflicted,
        } = entries.insert_id(tables, entry.id, index)?
        else {
            continue;
        };
        entries.reject(tables, first)?;
        entries.reject(tables, index)?;
        if newly_conflicted {
            emit(
                reporter,
                ValidationReason::StructureInvalid,
                ValidationObject::StructureDictionary,
                Some(entry.leaf_page),
            )?;
        }
    }
    Ok(())
}

fn report_outcomes<S: RecoverySink>(
    entries: &StructureIndex,
    tables: &Tables,
    reporter: &mut Reporter<'_, S>,
) -> Result<()> {
    let mut accepted = 0u64;
    let mut rejected = 0u64;
    for index in 0..entries.records_len() {
        if entries.record(tables, index)?.rejected {
            rejected = rejected.checked_add(1).ok_or(Error::ArithmeticOverflow(
                "recovery structure entries rejected",
            ))?;
        } else {
            accepted = accepted.checked_add(1).ok_or(Error::ArithmeticOverflow(
                "recovery structure entries accepted",
            ))?;
        }
    }
    reporter.structure_accepted(accepted)?;
    reporter.structure_rejected(rejected)
}

File: v4/go/internal/recovery/structure_index.go
package recovery

// Recovery of authoritative structured payloads by source-local ID
// (Rust recovery/structure_index.rs): the dense table-tree scan over
// the page-ownership set counts and recovers the structure records,
// the ID conflicts reconcile, and the outcomes fold into the report.

import (
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/live"
	"github.com/firehol/iprange/v4/go/internal/mapping"
	"github.com/firehol/iprange/v4/go/internal/validation"
)

// structureTableEvents is one structure table scan sink (Rust
// TableEvents).
type structureTableEvents interface {
	pageAccepted() error
	pageRejected(ioUnreadable bool) error
	unknown(reason validation.ValidationReason, page *uint32) error
	leaf(page uint32, expectedID uint64, cell []byte) error
}

// countStructureRecords counts the recovery-readable structure records
// of the network-enrichment kind (Rust structure_index::count: the
// table scan with the id bound and the payload digest proof).
func countStructureRecords(m *mapping.Mapping, meta format.Meta, pages *pageSet, check func() error) (uint64, error) {
	events := &structureCountEvents{meta: meta}
	if err := scanStructureTable(m, meta, meta.StructureIDRoot, pages, check, events); err != nil {
		return 0, err
	}
	return events.count, nil
}

// structureCountEvents counts the readable structure records (Rust
// Counter).
type structureCountEvents struct {
	meta  format.Meta
	count uint64
}

func (e *structureCountEvents) pageAccepted() error { return nil }
func (e *structureCountEvents) pageRejected(ioUnreadable bool) error {
	return nil
}
func (e *structureCountEvents) unknown(reason validation.ValidationReason, page *uint32) error {
	return nil
}
func (e *structureCountEvents) leaf(page uint32, expectedID uint64, cell []byte) error {
	record, err := format.DecodeStructureRecord(cell)
	if err != nil || uint64(record.ID) != expectedID || expectedID >= e.meta.StructureIDLimit {
		return nil
	}
	digest, err := format.StructurePayloadDigest(format.StructureKindNetworkEnrichmentV1, record.Payload)
	if err != nil {
		return err
	}
	if digest != record.Digest {
		return nil
	}
	next := e.count + 1
	if next == 0 {
		return overflowError("recovery structure count")
	}
	e.count = next
	return nil
}

// recoverStructureRecords reconciles the structure dictionary of one
// source (Rust structure_index::recover: the kind proof, the table
// scan with the membership dependence, the ID conflict reconciliation,
// and the outcome report).
func recoverStructureRecords(m *mapping.Mapping, meta format.Meta, memberships *membershipIndex, pages *pageSet, tables *tableStore, check func() error, rep *reporter) (*structureIndex, error) {
	if meta.StructureKind != format.StructureKindNetworkEnrichmentV1 {
		return nil, &format.Error{Code: format.CodeUnsupportedStructure, Detail: "recovery structure kind is unsupported"}
	}
	entries := newStructureIndex(tables, format.StructureKindNetworkEnrichmentV1)
	events := &structureRecoverEvents{
		meta:        meta,
		rep:         rep,
		memberships: memberships,
		entries:     entries,
		tables:      tables,
	}
	if err := scanStructureTable(m, meta, meta.StructureIDRoot, pages, check, events); err != nil {
		return nil, err
	}
	if err := reconcileStructureIDs(entries, tables, rep); err != nil {
		return nil, err
	}
	return entries, reportStructureOutcomes(entries, tables, rep)
}

// structureRecoverEvents wires the structure table scan into the
// reporter and the locator table (Rust Events: every leaf cell counts
// one examined record, undecodable records reject, the id bound and
// the payload digest reject with their exact classes, and the
// membership dependence rejects with the membership-invalid class).
type structureRecoverEvents struct {
	meta        format.Meta
	rep         *reporter
	memberships *membershipIndex
	entries     *structureIndex
	tables      *tableStore
}

func (e *structureRecoverEvents) pageAccepted() error {
	return e.rep.pageAccepted()
}

func (e *structureRecoverEvents) pageRejected(ioUnreadable bool) error {
	return e.rep.pageRejected(ioUnreadable)
}

func (e *structureRecoverEvents) unknown(reason validation.ValidationReason, page *uint32) error {
	return e.rep.emitPageUnknown(reason, validation.ObjectStructureDictionary, page)
}

func (e *structureRecoverEvents) leaf(page uint32, expectedID uint64, cell []byte) error {
	if err := e.rep.structureExamined(); err != nil {
		return err
	}
	record, err := format.DecodeStructureRecord(cell)
	if err != nil {
		return e.rep.structureRejected(1)
	}
	// The implied-slot id compare is its own refusal with the
	// StructureInvalid envelope (Rust structure_index.rs Events::leaf:
	// decode_record, then the id and limit proof).
	if uint64(record.ID) != expectedID || expectedID >= e.meta.StructureIDLimit {
		if err := e.rep.structureRejected(1); err != nil {
			return err
		}
		return e.unknown(validation.ReasonStructureInvalid, &page)
	}
	digest, err := format.StructurePayloadDigest(format.StructureKindNetworkEnrichmentV1, record.Payload)
	if err != nil {
		return err
	}
	if digest != record.Digest {
		if err := e.rep.structureRejected(1); err != nil {
			return err
		}
		return e.unknown(validation.ReasonStructureHashInvalid, &page)
	}
	membershipID := format.U32(record.Payload[24:28])
	rejected := false
	if membershipID != 0 {
		_, found, err := e.memberships.get(e.tables, membershipID)
		if err != nil {
			return err
		}
		rejected = !found
		if rejected {
			if err := e.unknown(validation.ReasonStructureMembershipInvalid, &page); err != nil {
				return err
			}
		}
	}
	var payload [format.NetworkEnrichmentV1PayloadSize]byte
	copy(payload[:], record.Payload)
	return e.entries.push(e.tables, structureLocator{
		id:           record.ID,
		membershipID: membershipID,
		leafPage:     page,
		payload:      payload,
		payloadLen:   format.NetworkEnrichmentV1PayloadSize,
		rejected:     rejected,
	})
}

// reconcileStructureIDs folds the source-ID registration conflicts
// (Rust reconcile_ids: both records of one conflicting id reject, and
// the first conflict streams the StructureInvalid envelope).
func reconcileStructureIDs(entries *structureIndex, tables *tableStore, rep *reporter) error {
	for index := uint64(0); index < entries.recordsLen(); index++ {
		entry, err := entries.record(tables, index)
		if err != nil {
			return err
		}
		insert, err := entries.insertID(tables, entry.id, index)
		if err != nil {
			return err
		}
		if !insert.duplicate {
			continue
		}
		if err := entries.reject(tables, insert.first); err != nil {
			return err
		}
		if err := entries.reject(tables, index); err != nil {
			return err
		}
		if insert.newlyConflicted {
			if err := rep.emitPageUnknown(validation.ReasonStructureInvalid, validation.ObjectStructureDictionary, &entry.leafPage); err != nil {
				return err
			}
		}
	}
	return nil
}

// reportStructureOutcomes folds the structure proof (Rust
// report_outcomes: the accepted and rejected counts).
func reportStructureOutcomes(entries *structureIndex, tables *tableStore, rep *reporter) error {
	var accepted, rejected uint64
	for index := uint64(0); index < entries.recordsLen(); index++ {
		entry, err := entries.record(tables, index)
		if err != nil {
			return err
		}
		if entry.rejected {
			rejected++
		} else {
			accepted++
		}
	}
	if err := rep.structureAccepted(accepted); err != nil {
		return err
	}
	return rep.structureRejected(rejected)
}

// scanStructureTable walks one dense structure table tree (Rust
// structure_index::scan: the root level from the id limit, then the
// node walk).
func scanStructureTable(m *mapping.Mapping, meta format.Meta, root uint32, pages *pageSet, check func() error, events structureTableEvents) error {
	if root == 0 {
		return nil
	}
	level, ok := format.StructureRootLevel(meta.StructureIDLimit)
	if !ok {
		return corruptError("recovery structure id limit is invalid")
	}
	var path [format.StructureTableMaxLevel + 1]uint32
	return scanStructureNode(m, meta, root, uint16(level), 0, &path, 0, pages, check, events)
}

// scanStructureNode walks one structure table node (Rust scan_node:
// the claim, the checked page, the dense header, the reserved-zero
// proof, and the leaf or directory walk).
func scanStructureNode(m *mapping.Mapping, meta format.Meta, pageNumber uint32, expectedLevel uint16, base uint64, path *[format.StructureTableMaxLevel + 1]uint32, depth int, pages *pageSet, check func() error, events structureTableEvents) error {
	if err := live.Checkpoint(check); err != nil {
		return err
	}
	claimed, reason, err := pages.claim(pageNumber, meta.PageCount, path[:], depth)
	if err != nil {
		return err
	}
	if !claimed {
		pageCopy := pageNumber
		return events.unknown(reason, &pageCopy)
	}
	page, problem := checkedPage(m, pageNumber, meta.PageCount)
	if problem != nil {
		return rejectStructurePage(events, pageNumber, problem.reason, problem.ioUnreadable)
	}
	header, headerProblem := format.InspectStructureTableHeader(page, meta.TxnID, uint32(format.StructureKindNetworkEnrichmentV1), &expectedLevel)
	if headerProblem != format.TreeHeaderProblemNone {
		return rejectStructurePage(events, pageNumber, structureHeaderReason(headerProblem), false)
	}
	end := uint16(format.StructureLeafEnd)
	if header.Level != 0 {
		end = uint16(format.StructureBranchEnd)
	}
	if !format.AllZero(page[end:]) {
		return rejectStructurePage(events, pageNumber, validation.ReasonPageReservedNonzero, false)
	}
	if err := events.pageAccepted(); err != nil {
		return err
	}
	if header.Level == 0 {
		return scanStructureLeaf(m, meta, page, header, pageNumber, base, pages, check, events)
	}
	return scanStructureBranch(m, meta, page, header, pageNumber, base, path, depth, pages, check, events)
}

// rejectStructurePage streams one refused structure page (Rust
// reject_page: the page-rejected count then the unknown envelope).
func rejectStructurePage(events structureTableEvents, pageNumber uint32, reason validation.ValidationReason, ioUnreadable bool) error {
	if err := events.pageRejected(ioUnreadable); err != nil {
		return err
	}
	return events.unknown(reason, &pageNumber)
}

// structureHeaderReason maps one dense header problem to its envelope
// class (Rust header_reason).
func structureHeaderReason(problem format.TreeHeaderProblem) validation.ValidationReason {
	switch problem {
	case format.TreeHeaderProblemBorn:
		return validation.ReasonPageBornTxnInvalid
	case format.TreeHeaderProblemType:
		return validation.ReasonPageTypeMismatch
	case format.TreeHeaderProblemLevel:
		return validation.ReasonTreeLevelInvalid
	default:
		return validation.ReasonPageHeaderInvalid
	}
}

// scanStructureLeaf walks one record page (Rust scan_leaf: the fixed
// record array, the all-zero skip, the slot-implied id, and the
// item-count proof).
func scanStructureLeaf(m *mapping.Mapping, meta format.Meta, page []byte, header format.PageHeader, pageNumber uint32, base uint64, pages *pageSet, check func() error, events structureTableEvents) error {
	found := 0
	for slot := 0; slot < format.StructureRecordSlots; slot++ {
		if err := live.Checkpoint(check); err != nil {
			return err
		}
		at := 32 + slot*format.StructureRecordSize
		cell := page[at : at+format.StructureRecordSize]
		if format.AllZero(cell) {
			continue
		}
		found++
		if err := events.leaf(pageNumber, base+uint64(slot), cell); err != nil {
			return err
		}
	}
	if found != int(header.ItemCount) {
		pageCopy := pageNumber
		return events.unknown(validation.ReasonPageHeaderInvalid, &pageCopy)
	}
	return nil
}

// scanStructureBranch walks one directory page (Rust scan_branch: the
// fixed child array, the coverage-scaled child base, and the
// item-count proof).
func scanStructureBranch(m *mapping.Mapping, meta format.Meta, page []byte, header format.PageHeader, pageNumber uint32, base uint64, path *[format.StructureTableMaxLevel + 1]uint32, depth int, pages *pageSet, check func() error, events structureTableEvents) error {
	span, ok := format.StructureSpanOfLevel(uint32(header.Level) - 1)
	if !ok {
		return overflowError("recovery structure coverage")
	}
	found := 0
	for index := 0; index < format.StructureDirectoryChildCount; index++ {
		if err := live.Checkpoint(check); err != nil {
			return err
		}
		child := format.U32(page[32+index*4:])
		if child == 0 {
			continue
		}
		found++
		if child < 2 || uint64(child) >= meta.PageCount {
			if err := events.unknown(validation.ReasonPageOutOfBounds, &child); err != nil {
				return err
			}
			continue
		}
		scaled, ok := checkedMul(span, uint64(index))
		if !ok {
			return overflowError("recovery structure coverage")
		}
		childBase, ok := checkedAdd(base, scaled)
		if !ok {
			return overflowError("recovery structure coverage")
		}
		if err := scanStructureNode(m, meta, child, header.Level-1, childBase, path, depth+1, pages, check, events); err != nil {
			return err
		}
	}
	if found != int(header.ItemCount) {
		pageCopy := pageNumber
		return events.unknown(validation.ReasonPageHeaderInvalid, &pageCopy)
	}
	return nil
}
